### PR TITLE
Add ability to report losses

### PR DESCRIPTION
**Pull request by fubuloubu**
fixes: #49

TODO:
- [x] tests
- [ ] docs

### CODE REVIEW

File: contracts/Vault.vy [hunk 1]
```diff
@@ -1166,20 +1168,35 @@ def report(_return: uint256) -> uint256:
...
+        last report, and should be accounted for on the Vault's balance sheet
     @return Amount of debt outstanding (iff totalDebt > debtLimit).
     """
 
     # Only approved strategies can call this function
     assert self.strategies[msg.sender].activation > 0
 
+    # We have a loss to report, do it before the rest of the calculations
+    if _loss > 0:
+        loss: uint256 = min(_loss, self.strategies[msg.sender].totalDebt)
```
Comment (dmolina79) [inline contracts/Vault.vy:1190]: gain and loss are supposed to be mutually exclusive? meaning if i have a loss to report there shouldn't be any gain on that call, unless im missing something.  

Just thinking if we need to require if loss > 0 gain should equal 0 and vice versa, only issue is that the strategy could mess up and get stuck failing to not report

File: contracts/Vault.vy [hunk 2]
```diff
@@ -1228,10 +1245,10 @@ def report(_return: uint256) -> uint256:
     # the credit increase we are offering (if any)
     # NOTE: This is just used to adjust the balance of tokens between the Strategy and
     #       the Vault based on the Strategy's debt limit (as well as the Vault's).
-    if _return < credit:  # credit surplus, give to Strategy
-        assert self.token.transfer(msg.sender, credit - _return)
-    elif _return > credit:  # credit deficit, take from Strategy
-        assert self.token.transferFrom(msg.sender, self, _return - credit)
+    if _gain < credit:  # credit surplus, give to Strategy
+        assert self.token.transfer(msg.sender, credit - _gain)
```
Comment (dmolina79) [inline contracts/Vault.vy:1249]: `_gain` could be zero, so we would give back completed credit amount right?

File: contracts/Vault.vy (hunk 1 above)
Comment (fubuloubu) [reply contracts/Vault.vy:1190]: I left it open here. I could see a scenario where you had both a gain and a loss to report e.g. you cashed in some reward tokens, but you lost some amount of underlying in the process. Of course, you could just mix the two together, but I think leaving it open makes it more flexible.

File: contracts/Vault.vy (hunk 2 above)
Comment (fubuloubu) [reply contracts/Vault.vy:1249]: Yes, that's the idea

File: contracts/Vault.vy [hunk 3]
```diff
@@ -1273,7 +1290,7 @@ def report(_return: uint256) -> uint256:
 
     log StrategyReported(
         msg.sender,
-        _return,
+        _gain,
```
Comment (dmolina79) [inline contracts/Vault.vy:1293]: do we want to track reported loss value in event as well?

**Review by dmolina79 — APPROVED** [review]
LGTM!, recommend a second review from @banteg since he requested the change

File: contracts/Vault.vy (hunk 3 above)
Comment (fubuloubu) [reply contracts/Vault.vy:1293]: Yes, I'll probably change up the events in a later PR too

File: contracts/Vault.vy [hunk 4]
```diff
@@ -1155,7 +1157,7 @@ def expectedReturn(_strategy: address = msg.sender) -> uint256:
 
 
 @external
-def report(_return: uint256) -> uint256:
+def report(_gain: uint256, _loss: uint256) -> uint256:
```
Comment (dmolina79) [inline contracts/Vault.vy:1160]: is it a wash if gain and loss are reported by any chance with same number?

Comment (fubuloubu) [reply contracts/Vault.vy:1162]: A reported loss gets accounted for first, reducing debt, so it'll look like you earned more with less I suppose.

File: contracts/Vault.vy [hunk 5]
```diff
@@ -102,10 +103,12 @@ event StrategyAdded:
...
-    returnAdded: uint256
-    debtAdded: uint256
-    totalReturn: uint256
+    gain: uint256
+    loss: uint256
+    totalGain: uint256
+    totalLoss: uint256
     totalDebt: uint256
+    debtAdded: uint256
     debtLimit: uint256
```
Comment (fubuloubu) [inline contracts/Vault.vy:112]: @banteg made a big change here

File: tests/functional/vault/test_losses.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    token.approve(vault, 500, {"from": gov})
+    vault.deposit(500, {"from": gov})
+
+    strategy.harvest({"from": gov})
+    assert token.balanceOf(strategy) == 500
+
+    # First loss
+    strategy._takeFunds(100, {"from": gov})
+    strategy.harvest({"from": gov})
+    params = vault.strategies(strategy).dict()
```
Comment (fubuloubu) [inline tests/functional/vault/test_losses.py:28]: We have to do this to all the tests :smiling_imp:

Comment (fubuloubu) [reply tests/functional/vault/test_losses.py:28]: https://github.com/iearn-finance/yearn-vaults/pull/76

File: tests/functional/vault/test_strategies.py [hunk 7]
```diff
@@ -20,7 +20,7 @@ def test_addStrategy(web3, gov, vault, strategy, rando):
     with brownie.reverts():
         vault.addStrategy(strategy, 1000, 10, 50, {"from": rando})
 
-    assert vault.strategies(strategy) == [0, 0, 0, 0, 0, 0, 0]
+    assert vault.strategies(strategy) == [0, 0, 0, 0, 0, 0, 0, 0]
```
Comment (banteg) [inline tests/functional/vault/test_strategies.py:23]: With so many params, should we switch to comparing dicts?

**Review by banteg — APPROVED** [review]
Looks great, this opens up a whole new world of possibilities!

File: tests/functional/vault/test_strategies.py (hunk 7 above)
Comment (fubuloubu) [reply tests/functional/vault/test_strategies.py:23]: See #76 (which I didn't know existed until today, lol)

File: contracts/Vault.vy (hunk 4 above)
Comment (dmolina79) [reply contracts/Vault.vy:1162]: makes sense, would it be good to add a test case to check values after side effects?

Comment (fubuloubu) [reply contracts/Vault.vy:1162]: We added a test case in 1cb83d1, do you think it can be improved?

Comment (dmolina79) [reply contracts/Vault.vy:1162]: yeah, maybe we can add a mock strategy that sends bad signals in testing like gain=100K and loss=100K just to check the state of the vault @Macarse, would that make sense?

Comment (dmolina79) [reply contracts/Vault.vy:1162]: Ill add it as an issue to track to improve testing, just for completion sake i would say, glad to take a stab at it

File: contracts/Vault.vy [hunk 8]
```diff
@@ -1155,7 +1159,7 @@ def expectedReturn(_strategy: address = msg.sender) -> uint256:
 
 
 @external
-def report(_return: uint256) -> uint256:
+def report(_gain: uint256, _loss: uint256) -> uint256:
```
Comment (Grandthrax) [inline contracts/Vault.vy:1162]: Gain is the wrong word to use. Gain implies that the number is profit. But that is not the case. The number is profit + debtOutstanding

File: contracts/Vault.vy [hunk 9]
```diff
@@ -1166,20 +1170,35 @@ def report(_return: uint256) -> uint256:
...
         reserves. Anything outside of those bounds is abnormal behavior.
 
         All approved strategies must have increased diligence around
         calling this function, as abnormal behavior could become catastrophic.
-    @param _return
-        Amount Strategy has made on it's investment since its last report,
-        and is free to be given back to Vault as earnings
+    @param _gain
+        Amount Strategy has realized as a gain on it's investment since its
+        last report, and is free to be given back to Vault as earnings
```
Comment (Grandthrax) [inline contracts/Vault.vy:1180]: Plus any debt outstanding

File: contracts/Vault.vy [hunk 10]
```diff
@@ -1166,20 +1170,35 @@ def report(_return: uint256) -> uint256:
...
     """
 
     # Only approved strategies can call this function
     assert self.strategies[msg.sender].activation > 0
 
+    # We have a loss to report, do it before the rest of the calculations
+    if _loss > 0:
+        loss: uint256 = min(_loss, self.strategies[msg.sender].totalDebt)
+        self.strategies[msg.sender].totalLoss += loss
+        self.strategies[msg.sender].totalDebt -= loss
```
Comment (Grandthrax) [inline contracts/Vault.vy:1194]: You either need to decrease from totalGain as well or change the _expectedReturn method

File: contracts/Vault.vy [hunk 11]
```diff
@@ -1124,14 +1128,14 @@ def creditAvailable(_strategy: address = msg.sender) -> uint256:
 def _expectedReturn(_strategy: address) -> uint256:
```
Comment (Grandthrax) [inline contracts/Vault.vy:1128]: if totalGain does not include losses then you need to update this method. See my comment further down where you calculate loss in report() method

File: contracts/Vault.vy (hunk 8 above)
Comment (Grandthrax) [inline contracts/Vault.vy:1162]: Please can we have a second report method with 1 input that calls this with loss = 0. So that this vault upgrade will still work with legacy strategies

**Review by Grandthrax — COMMENTED** [review]
mostly looks good. Some clarification should be added to show the user how to deal with debt outstanding.

File: contracts/Vault.vy (hunk 8 above)
Comment (fubuloubu) [reply contracts/Vault.vy:1162]: You can't migrate a strategy to a newer Vault, the solution here will be to upgrade and re-release the strategy.

Comment (fubuloubu) [reply contracts/Vault.vy:1162]: perhaps this should be revisited, have it be simply reported gains, but then add a `debtPayment` field too

Comment (Grandthrax) [reply contracts/Vault.vy:1162]: That would certainly make it easier for new users to understand.
